### DISCUSSION TITLE
How to bail out early with expired refresh token using exchange-auth  ?

### DISCUSSION THREAD

**Discussion by vinamelody** [urql-graphql/urql#1239] [Q&A]
Hi, I am trying to resolve an issue where i;m using Refresh Token Rotation and the refresh token has expired. When it happens, my idea is to use `auth0.loginWithRedirect` to get a fresh, new authState. I am not storing token in localStorage / sessionStorage manually.

I read this explanation here <https://github.com/FormidableLabs/urql/pull/939> but still have more doubts.

1. Why does the `operation` is returned even when there is no authState in `addAuthToOperation` ?
2. Why urql still execute api calls when `willAuthError` returns true ?
3. How to bail out early and halt the API call?

```javascript
const urqlClient = createClient({
  url: `${Constants.API_ENDPOINTS.base}${Constants.API_ENDPOINTS.authGraphQL}`,
  requestPolicy: 'network-only',
  exchanges: [dedupExchange, cacheExchange,
      authExchange({
        addAuthToOperation: ({ authState, operation }) => {
          if (!authState || !authState.token) {
            return operation;
          }
          if (isExpired(authState.token)) {
            console.log('token is expired')
            return operation
          }
          const fetchOptions =
          typeof operation.context.fetchOptions === 'function'
          ? operation.context.fetchOptions()
          : operation.context.fetchOptions || {}
          
          const op = makeOperation(
            operation.kind,
            operation,
            {
              ...operation.context,
              fetchOptions: {
                ...fetchOptions,
                headers: {
                  ...fetchOptions.headers,
                  "Authorization": `Bearer ${authState.token}`,
                  "Accept": 'application/json'
                },
              },
            },
          )
          return op
        },
        willAuthError: ({ operation, authState}) => {
          if (!authState) return true
          if (isExpired(authState.token)) {
            console.log('willAuthError token is expired, how to bail out early and not sending out the API call')
            return true
          }
          return false
        },
        didAuthError: async ({ error }) => {
          // will get "CombinedError", message: "[Network] Bad Request" here because API call went out without Authorization
          console.log('didAuthError', error)
          return error.response.status === 401
        },
        getAuth: async ({ authState, mutate }) => {
          if (!authState) {
            try{
              const token = await auth0Client.getTokenSilently()
              return { token }
            } catch (error) {
              console.log('getAuth fails to renew token error', error)
              // last resort, I'm redirecting them to menu and it will be an auto login
              await auth0LoginWithRedirect(window.location.pathname)
              return null
            }
          }
        }
      }),
      fetchExchange],
  })
```

**Comment by kitten** [comment]
- `willAuthError` will block operations when it's known that the `authState` has expired. It's a way to signal that future operations won't succeed.
- `addAuthToOperation` is always called for each operation. It's safe there not to check for expiration and just to pass the authentication state (e.g. a header) onto the operation
- it shouldn't execute any operation until `getAuth` completes with a new authentication state once `willAuthError` or `didAuthError` have triggered once

**Reply by vinamelody** [reply]
Hi @kitten in my tries, even though `willAuthError` returns true, operation was still sent out ... is there anything that I missed?

**Reply by kitten** [reply]
From a quick look at the snippet again it seems that your `getAuth` function does nothing after initialisation, so that could be the reason why it's not working. You're basically instructing it to never refresh 😅

**Reply by vinamelody** [reply]
oh ok hmm i didn't see that perspective before 🧐, let me try and see if that's really the reason

**Reply by vinamelody** [reply]
Hi @kitten I modify it to be like below code and it's now able to refresh the token. But there's one more negative scenario i need to handle.

When the token refresh failed and the error is caught inside `getAuth`, auth0 returns `Error: Unknown or invalid refresh token` (error 403)
 
Thus, inside `getAuth`, I will return `{ null }` -->  API call was sent out without `Authorization` header --> `didAuthError` produces combined error 400 Bad request (hence it's not 401 or 403 anymore)

I guess, what i want to achieve here is for the UI to recognize that when the refresh token failed, I want to offer the use an action called `loginWithRedirect`... 

there's definitely one more thing i missed 😅 but i couldn't figure out ...

```javascript
const urqlClient = createClient({
  url: `${Constants.API_ENDPOINTS.base}${Constants.API_ENDPOINTS.authGraphQL}`,
  requestPolicy: 'network-only',
  exchanges: [dedupExchange, cacheExchange,
      ErrorExchange({
        onError: ({ error }) => {
          console.log('errorExchange', error)

        }
      }),
      authExchange({
        addAuthToOperation: ({ authState, operation }) => {
          if (!authState || !authState.token) {
            return operation;
          }
          const fetchOptions =
          typeof operation.context.fetchOptions === 'function'
          ? operation.context.fetchOptions()
          : operation.context.fetchOptions || {}
          
          const op = makeOperation(
            operation.kind,
            operation,
            {
              ...operation.context,
              fetchOptions: {
                ...fetchOptions,
                headers: {
                  ...fetchOptions.headers,
                  "Authorization": `Bearer ${authState.token}`,
                  "Accept": 'application/json'
                },
              },
            },
          )
          return op
        },
        willAuthError: ({ operation, authState}) => {
          if (!authState) return true
          if (isExpired(authState.token)) {
            console.log('willAuthError token is expired, how to bail out early and not sending out the API call')
            return true
          }
          return false
        },
        didAuthError: async ({ error }) => {
          // will get "CombinedError", message: "[Network] Bad Request" here because API call went out without Authorization
          console.log('didAuthError', error)

          // 401 might be login_required
          // 403 might be Invalid Refresh Token
          return error.response.status === 401 || error.response.status === 403
        },
        getAuth: async ({ authState, mutate }) => {
          if (!authState) {
            try{
              const token = await auth0Client.getTokenSilently()
              return { token }
            } catch (error) {
              console.log('getAuth fails to renew token error', error)
              // last resort, I need to return an error              
              return null
            }
          }
        }
      }),
      fetchExchange],
  })

**Reply by kitten** [reply]
I'm not quite sure what you mean, but the docs suggest the use cases f the `errorExchange` to introduce side effects

**Reply by vinamelody** [reply]
sorry, i meant when the refresh token failed, the error being output is from
```
catch (error) {
              console.log('getAuth fails to renew token error', error)
              // last resort, I need to return an error              
              return null
            }
```
hence this returns the `operation` causing API calls getting 400 Bad request
```
addAuthToOperation: ({ authState, operation }) => {
          if (!authState || !authState.token) {
            return operation;
          }
```

**Reply by kitten** [reply]
Yep, you don't need to catch errors in `getAuth` in fact, when you don't the auth error will fall through to the `errorExchange`.

Also the `authExchange` as detailed in a later part around errors will only attempt authentication once then it'll back off.

You can handle special logout logic either in your `getAuth` catch or in `errorExchange`.

So the repeated failure after `getAuth` fails is safe. You can let the result's error fall through and trust that once it reaches `errorExchange` the `authExchange` has tried to reauthenticate. So there the error means that the user is unauthenticated.

**Reply by vinamelody** [reply]
hmm... when i remove the catch, console.log prints this error, still, not passing it to errorExchange 🤔

```
Uncaught (in promise) TypeError: Cannot read property 'kind' of null
    at Q (urql-core.mjs:288)
    at wonka.mjs:369
    at wonka.mjs:660
    at Array.forEach (<anonymous>)
    at b (wonka.mjs:659)
    at b (wonka.mjs:119)
    at wonka.mjs:547
    at b (wonka.mjs:119)
    at wonka.mjs:533
    at Array.forEach (<anonymous>)
    at next (wonka.mjs:532)
    at Map.forEach (<anonymous>)
    at m (urql-exchange-auth.mjs:68)
    at B (urql-exchange-auth.mjs:82)
```

the code:
```
getAuth: async ({ authState, mutate }) => {
          if (!authState) {
            const token = await auth0Client.getTokenSilently()
            console.log('initial token', token)
            isTokenExpired(token)
            return { token }
          }

          const newToken = await auth0Client.getTokenSilently()
          if (newToken) {
            console.log('newToken', newToken)
            return { token: newToken }
          }
          
          return null
        }
      }),
```

tried with just `return { token: await auth0Client.getTokenSilently() }` too

**Reply by vinamelody** [reply]
>You can handle special logout logic either in your getAuth catch or in errorExchange.

I want to return a new form of error to the UI such as this, and the UI will offer the manual relogin for the user.

```
errObj.errors = [{
              error: err.error,
              status: 401,
              error_description: 'Your session has expired, please login again.'
            }]
```

**Reply by kitten** [reply]
An error like `kind of null` sounds like `addAuthOperation` may be returning `null` in this case 🤔 either way, I believe after taking another look either should work. So in either case you'll see `getAuth` being triggered, erroring`, another attempt at sending the operation, and then the error will fall through 

Now, the `errorExchange` doesn't allow you to remap an error and you'd have to write a custom exchange for that. That being said it may not make much sense to rewrite the message and error. That's UI logic and not business logic and you'd be adding a layer of indirection.

The `errorExchange` is meant for side-effects like changing some state to replace the UI with an error, have a passive error pop up, redirect to another page. For logout you probably want to take some more drastic measures.

However at this stage I do have to point out, I'm guessing 😅 the docs page for "Authentication" answers all of these questions with a step by step guide though.

**Reply by vinamelody** [reply]
if `willAuthError` returns true, is `addAuthOperation` still executed ?

**Reply by vinamelody** [reply]
I think here's the flow now:

1. initial token (wait until both token and refresh token expired)
2. after expired, try to hit the API
3. `willAuthError` returns true (because of the token expiry check)
4. go to `getAuth` and it tries to `getTokenSilently` but this will also fail as refresh token is also expired (hence auth0 403)
5. go to `addAuthOperation` when there's no token, returns `operation` -> hence API produces 400 Bad request

desired flow: 1-2-3-4 -> immediately forward auth error to errorExchange (don't go to 5)

**Reply by vinamelody** [reply]
![Screenshot 2020-12-21 at 9 29 27 AM](https://user-images.githubusercontent.com/990511/102730383-895df880-436f-11eb-8b18-bad530daa186.png)
adding the log for better visualisation 😅

**Comment by hongkheng** [comment]
To add on to @vinamelody's post, the authentication is handled by a 3rd party service (a different api). This urql flow assumes `operation` is handling both auth and api which is not the case for us. What we are trying to achieve is to retrieve the token refresh error thrown by the 3rd party service, but `operation` will be fired (we trying to see if we can bail out and not call operation) and the error returned by the urql will now be API's error (400 bad request) instead of the original error thrown by the auth error.

Will need some light on how we can thrown this auth error back to the UI instead.

**Reply by kitten** [reply]
Ah I see, so your authentication returns different information than your API.

Currently there's no special logic in the `authExchange` to handle this, but there's a clear reason for this and the tl;dr is that this doesn't fit well into GraphQL:

1. Each operation may be sent and will be responded to either from the cache or the server with a result
2. Each operation may be responded to with an error result as per the spec or the request may fail. In the case of complete failure we abstract this as `CombinedError#networkError`
3. In `authExchange` the refresh call is a side-effect with an unrelated error.

This is basically the problem you have which may come to an abstraction in authentication that you haven't mentioned here.

Basically refreshing an authentication method is a side-effect which you're performing in response to a 403 unauthenticated code. But an API could also return a GraphQLError to represent this with more information.

The refresh call is incidental, so it may indeed fail as well. At that point two things are important which I tried to highlight:

- we can't give you the second error of the refresh call, since it's a random call that may be called _once_ for _multiple operations_. Furthermore if we used this error then the original results errors may be lost.
- Your API should possibly already deliver enough information on the original "403-like response"

So this runs really deep but what I'm trying to say is that if I get this correctly you know what you want, I hope I understand where you want to get to but the problem from my end is that I don't think that's a road that you want to go down to. And if you did I'd have to recommend you to write a custom exchange 😅

So all in all, if the API is under your control then treat the error from the GraphQL API as one that can return rich error info. Maybe by using a full GraphQLError in it's response

**Comment by hongkheng** [comment]
Thanks for your reply.

The API is not under my control, unfortunately I don't even want the API to even get called. I'm not interested in the auth error from the API in the case when the refresh token has expired because it is not useful. What I'm trying to do is handling all the authentication logic as much as in `authExchange` if possible. 

e.g When refresh token has expired, we want to throw that error to the UI and let the user knows and initiate a new user login. This is quite challenging now as the `CombinedError#networkError` is returning from the API which is not useful. We want the auth error to be thrown up instead. 

I'm trying another method to return a custom operation to return with this auth error, but I reckon this would not be possible. 

```js
addAuthToOperation: ({ authState, operation }) => {
     // if authState has the error
     // return an empty operation
     if (authState && authState.authError) {
       return makeOperation('teardown', operation)
     }
},
getAuth: async ({ authState, mutate }) => {
    console.log('getAuth', authState)
    if (!authState) {
     // ... refresh to get the token
    }
          
    // try to get a new refresh token
    try {
      const newToken = await auth0Client.getTokenSilently()
      if (newToken) {
         console.log('newToken', newToken)
         return { token: newToken }
       }
     } catch (error) {
        // return error in the authState 
        return { error: error, authError: 'auth0Error'}
     }
     return null
}
```

**Reply by kitten** [reply]
> When refresh token has expired, we want to throw that error to the UI and let the user knows and initiate a new user login.

I suppose it all comes down whether it matters which error you throw up. Either way, the result's error will be thrown up as long as an expired / non-existing token will also end up erroring out on the API. But the caught error as you're showing here will never be thrown up. This ultimately shouldn't matter as the flow will be: `operation - (willAuthError?) - (getAuth) -> retry operation -> error (thrown upwards)` or `operation -> result with auth error -> (didAuthError?) - (getAuth) -> retry operation -> error (thrown upwards)`

**Reply by vinamelody** [reply]
The tricky thing is that we have 2 separate endpoints, auth0 and app gateway 😅
both are beyond our control.
so, somehow the original auth error becomes replaced by server error

**Reply by kitten** [reply]
I see what you mean; well, like I meant, the refresh error in `getAuth` doesn't quite have anything to do with the prior or following operations and their results, but would it be possible for you to catch the error and keep it around in a local variable or sth of the likes until the `errorExchange` triggers? Because to be frank, that seems like the most obvious solution

**Reply by vinamelody** [reply]
alright that seems to be the only direction for now :) thank you soo much for your help 🙇‍♀️
